subscriptions/providers/apple_in_app/__init__.py

```diff
@@ -0,0 +1,189 @@
+        raise AppleInvalidOperation()
+
+    def webhook(self, request: Request, payload: dict) -> Response:
+        handlers = {
+            AppleReceiptRequest: self._handle_receipt,
+            AppleAppStoreNotification: self._handle_app_store,
+        }
+
+        for request_class, handler in handlers.items():
+            with suppress(ValidationError):
+                instance = request_class.parse_obj(payload)
+            if instance is not None:
+                return handler(request, instance)
+
+        # Invalid, unhandled request.
+        return Response(status=HTTP_400_BAD_REQUEST)
+
+    def check_payments(self, payments: Iterable[SubscriptionPayment]):
+        for payment in payments:
+            if payment.status != SubscriptionPayment.Status.COMPLETED:
+                # All the operations that we care about should be completed before they reach us.
+                raise AppleSubscriptionNotCompletedError(payment.provider_transaction_id)
+
+    @staticmethod
+    def _get_validated_in_app_product(response: AppleVerifyReceiptResponse) -> AppleInApp:
+        if not response.is_valid or response.receipt.bundle_id != settings.APPLE_BUNDLE_ID:
+            raise AppleReceiptValidationError()
+        return one(response.receipt.in_apps)
+
+    @transaction.atomic
+    def _handle_receipt(self, request: Request, payload: AppleReceiptRequest) -> Response:
+        receipt = payload.transaction_receipt
+
+        # Validate the receipt. Fetch the status and product.
+        receipt_data = self.api.fetch_receipt_data(receipt)
+        single_in_app = self._get_validated_in_app_product(receipt_data)
+
+        # Check whether this receipt is anyhow interesting:
+        with suppress(SubscriptionPayment.DoesNotExist):  # noqa (DoesNotExist seems not to be an exception for PyCharm)
+            payment = SubscriptionPayment.objects.get(provider_codename=self.codename,
+                                                      provider_transaction_id=single_in_app.transaction_id)
+            # We've already handled this. And all the messages coming to us from iOS should be AFTER
+            # the client money were removed.
+            return Response(
+                SubscriptionPaymentSerializer(payment).data,
+                status=HTTP_200_OK,
+            )
+
+        # Find the right plan to create subscription.
+        try:
+            search_kwargs = {
+                f'metadata__{self.product_id_metadata_field}': single_in_app.product_id
+            }
+            plan = Plan.objects.get(**search_kwargs)
+        except Plan.DoesNotExist:
+            return Response(status=HTTP_404_NOT_FOUND)
+
+        # Create subscription payment. Subscription is created automatically.
+        subscription_payment = SubscriptionPayment.objects.create(
+            provider_codename=self.codename,
+            provider_transaction_id=single_in_app.transaction_id,
+            status=SubscriptionPayment.Status.COMPLETED,
```

Comment:
You don't check `receipt_data`. Is it possible that receipt is in some "PENDING" state?

Reply:
I believe that they can't, but I'll make sure once again (maybe there is something that would allow for receipt generation before payment finalisation).